### PR TITLE
CT5UP-83: Add plugins and handle hot module reloading

### PR DESCRIPTION
**Pull request by NataliaMarkaki**

### CODE REVIEW

**Comment by briandipalma** [comment]
Has this been tested on SalesTrader or one of our other apps?

**Comment by briandipalma** [comment]
Conflicts in the PR.

**Comment by briandipalma** [comment]
Conflicts need to be addressed.

File: express-dev-server/webpack.js [hunk 1]
```diff
@@ -1,21 +1,55 @@
 const webpack = require("webpack");
 const webpackDevMiddleware = require("webpack-dev-middleware");
+const webpackHotMiddleware = require("webpack-hot-middleware");
+
+const { HMR } = require("../webpack-config-app/config");
+const isHotRealoadingActivated = HMR;
```
Comment (briandipalma) [inline express-dev-server/webpack.js]: This approach to enabling HMR in dev-server doesn't convince me. Changing a module's exported value seems odd/hacky to me. We should either pass in a flag or figure out we want to enable HMR from the webpack config. Maybe analyze the webpack `entry` value?

File: webpack-config-app/index.js [hunk 2]
```diff
@@ -20,16 +21,21 @@ const testsScriptRunning = basename(process.argv[1]) === "tests.js";
 const isBuild = buildScriptRunning || lifeCycleEvent === "build";
 const isTest = testsScriptRunning || lifeCycleEvent.startsWith("test");
 
-module.exports.webpackConfigGenerator = function webpackConfigGenerator({
-  basePath,
-  version = "dev",
-  i18nFileName = `i18n-${version}.js`,
-  uglifyOptions
-}) {
+const { HMR } = require("./config");
```
Comment (briandipalma) [inline webpack-config-app/index.js]: This should be a CLI argument like the `sourceMaps` option.

File: webpack-config-app/plugins.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/* eslint no-param-reassign: "off" */
+const webpack = require("webpack");
+const webpackVersion = require("../../webpack/package.json").version;
+
+module.exports = function configureBundlePlugins(
+  webpackConfig,
+  isHotRealoadingActivated
+) {
+  if (isHotRealoadingActivated) {
+    if (webpackVersion !== "1.15.0") {
```
Comment (briandipalma) [inline webpack-config-app/plugins.js]: This package use webpack 2 and above so this code block seems unnecessary?

File: express-dev-server/webpack.js [hunk 4]
```diff
@@ -1,9 +1,14 @@
...
+const webpackHotMiddleware = require("webpack-hot-middleware");
+const parseArgs = require("minimist");
+
+const { hot } = parseArgs(process.argv.slice(2));
 
 module.exports = (application, webpackConfig) => {
   const compiler = webpack(webpackConfig);
-  const devMiddlewareOptions = {
+
+  let devMiddlewareOptions = {
```
Comment (briandipalma) [inline express-dev-server/webpack.js]: Why did you change this to `let`? It looks like `const` should still work. If it does remove the added line to reduce the diff.

File: cli/templates/app-blank/src/index.js [hunk 5]
```diff
@@ -2,3 +2,7 @@ import Example from "./example/Example";
 import "./index.scss";
 
 console.log(new Example().sayHello());
+
+if (module.hot) {
```
Comment (briandipalma) [inline cli/templates/app-blank/src/index.js]: Just curious; will we need to add this code to any module we want to hot reload for it to work?

File: express-dev-server/webpack.js [hunk 6]
```diff
@@ -1,9 +1,14 @@
 const webpack = require("webpack");
 const webpackDevMiddleware = require("webpack-dev-middleware");
+const webpackHotMiddleware = require("webpack-hot-middleware");
+const parseArgs = require("minimist");
```
Comment (briandipalma) [inline express-dev-server/webpack.js]: `minimist` seems to be missing as a dependency in `package.json`?

File: express-dev-server/webpack.js (hunk 4 above)
Comment (NataliaMarkaki) [inline express-dev-server/webpack.js]: Oh that was a mistake. I needed it at first but after the changes you suggested it should be a const.

File: cli/templates/app-blank/src/index.js (hunk 5 above)
Comment (NataliaMarkaki) [inline cli/templates/app-blank/src/index.js]: In theory yes. Especially if we have react components. Although while testing it, I changed the output of sayHello() function and it changed instantly even if that branch doesn't exist in the Example class file.
